### ISSUE TITLE
Is it normal that every worker print the same loss and metric values when using multiworker distributed stragety?

### ISSUE THREAD

**Issue by AlexanderJLiu** [tensorflow/tensorflow#39343]
**Descriptions**

I've created a Keras model, and trained it with multiworker distributed strategy using `tf2.2.0`.

The training data I used is dataset gengerated from numpy array. And according to the [shard policy](https://www.tensorflow.org/api_docs/python/tf/data/experimental/DistributeOptions), the dataset will be auto sharded by `DATA policy`. That is to say every worker will handle different portion of the whole dataset.

After training, I notice that all the workers print the same loss and metric values just like they all trained with the identical data. But according to my knowledge, if every worker train with different data, the loss and metric values of them should be different. Is it because the shard policy didn't take effect?

I also know that multi-worker uses all-reduce communication method to keep variables in sync. So, the values printed are calculated after all-reduce? This could explain the same values maybe.

**Questions:**

1. Is this normal? And why?
2. How to check(or make sure) that different workers use different data portion for training(or the shard policy take effect)?
3. Are the values printed for all workers after all-reduce operation or just for that single worker?
4. When will all-reduce execute, on every batch end or epoch end?
5. When I use just numpy array for training, does the shard policy work? Or every worker will use the whole data for its training?
6. In general, dataset for every worker's fit method is same. When different, training will success as well, does shard policy work at this time?

**Code snippets:**

```python
class ThreeLayerMLP(keras.Model):
    def __init__(self, name=None):
        super().__init__(name=name)
        self.dense_1 = layers.Dense(64, activation='relu', name='dense_1')
        self.dense_2 = layers.Dense(64, activation='relu', name='dense_2')
        self.pred_layer = layers.Dense(10, name='predictions')

    def call(self, inputs):
        x = self.dense_1(inputs)
        x = self.dense_2(x)
        return self.pred_layer(x)


def main(argv):
    del argv  # Unused args
    strategy = tf.distribute.experimental.MultiWorkerMirroredStrategy()
    BATCH_SIZE_PER_REPLICA = 64
    BATCH_SIZE = BATCH_SIZE_PER_REPLICA * strategy.num_replicas_in_sync
    print('Number of devices: %d' % strategy.num_replicas_in_sync)

    with strategy.scope():
        model = model = ThreeLayerMLP(name='3_layer_mlp')
        model.compile(
            loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
            optimizer=keras.optimizers.RMSprop())

    log_dir = FLAGS.logs
    tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir=log_dir,
                                                          histogram_freq=1,
                                                          update_freq='batch')

    np.random.seed(0)
    x_train, y_train = (np.random.random(
        (60000, 784)), np.random.randint(10, size=(60000, 1)))
    x_test, y_test = (np.random.random(
        (10000, 784)), np.random.randint(10, size=(10000, 1)))

    # options = tf.data.Options()
    # options.experimental_distribute.auto_shard_policy = tf.data.experimental.AutoShardPolicy.DATA
    train_dataset = tf.data.Dataset.from_tensor_slices((x_train, y_train))
    train_dataset = train_dataset.shuffle(1024).batch(BATCH_SIZE)


    model.fit(
        train_dataset,
        epochs=5,
        steps_per_epoch=10,
        callbacks=tensorboard_callback)

    model_dir = FLAGS.logs + '/models/' + str(task_index)
    model.save(model_dir)


if __name__ == '__main__':
    app.run(main)
```

I'm cofused by these questions for long time. I will appreciate your help very much.

Thanks~

**Comment by yuefengz** [comment]
This is expected. The metric values would be allreduced in `fit` method. By default, `MultiWorkerMirroredStrategy` will shard your data across workers. You can use `auto_shard_policy` to turn off that behavior.
